src/promptflow-tools/promptflow/tools/common.py

```diff
@@ -442,6 +444,55 @@ def render_jinja_template(prompt, trim_blocks=True, keep_trailing_newline=True,
         raise JinjaTemplateError(message=error_message) from e
 
 
+def render_jinja_template_wrapper(prompt, trim_blocks=True, keep_trailing_newline=True, **kwargs):
+    # Use this dict to store the user written roles in case sensitive and its escaped value.
+    # We need this dict to unescape the role after the chat string is parsed.
+    escape_dict = {}
+    if isinstance(kwargs, dict):
+        updated_kwargs = {key: escape_roles(value, escape_dict) for key, value in kwargs.items()}
+    return (
+        render_jinja_template(
+            prompt,
+            trim_blocks=trim_blocks,
+            keep_trailing_newline=keep_trailing_newline,
+            **updated_kwargs,
+        ),
+        escape_dict,
+    )
+
+
+def escape_roles(val, escape_dict: dict):
+    """
+    Escape the roles in the prompt inputs to avoid the input string with pattern '# role' get parsed.
+    """
+    if not isinstance(val, str):
+        return val
+
+    pattern = r"(?i)^\s*#?\s*(" + "|".join(VALID_ROLES) + r")\s*:\s*\n"
+    roles = re.findall(pattern, val, flags=re.MULTILINE)
+    for role in roles:
+        if role not in escape_dict:
+            escape_dict[role] = str(uuid.uuid4())
```

Comment:
Why must use uuid? Any reason cannot use a constant hard code string like `$$<a-hash-of-the-reserved-tag>$$`?

Reply:
Even for uuid, it should be a fixed one, instead of we unescape all the uuids in input

Reply:
We cannot use a hard-coded hash str for each role, as the same role might be in various case formats. For example, the 'system' role may vary in input as 'system', 'System', 'SysteM','SYSTEM', etc. To convert the escaped roles back to the original str, we need to use different uuids for each case.